src/osdc/Objecter.cc

```diff
@@ -5072,58 +5033,62 @@ void Objecter::enumerate_objects(
   if (!osdmap->test_flag(CEPH_OSDMAP_SORTBITWISE)) {
     rl.unlock();
     lderr(cct) << __func__ << ": SORTBITWISE cluster flag not set" << dendl;
-    on_finish->complete(-EOPNOTSUPP);
+    std::move(on_finish)(boost::system::error_code(
+			   EOPNOTSUPP, boost::system::system_category()),
+      {}, {});
     return;
   }
   const pg_pool_t *p = osdmap->get_pg_pool(pool_id);
   if (!p) {
     lderr(cct) << __func__ << ": pool " << pool_id << " DNE in osd epoch "
 	       << osdmap->get_epoch() << dendl;
     rl.unlock();
-    on_finish->complete(-ENOENT);
+    std::move(on_finish)(boost::system::error_code(
+			   ENOENT, boost::system::system_category()),
+      {}, {});
     return;
   } else {
     rl.unlock();
   }
 
-  ldout(cct, 20) << __func__ << ": start=" << start << " end=" << end << dendl;
-
-  // Stash completion state
-  C_EnumerateReply *on_ack = new C_EnumerateReply(
-      this, next, result, end, pool_id, on_finish);
-
   ObjectOperation op;
   op.pg_nls(max, filter_bl, start, 0);
 
+  auto on_ack = std::make_unique<CB_EnumerateReply>(this, end, pool_id, std::move(on_finish));
+  auto epoch = &on_ack->epoch;
+  auto budget = &on_ack->budget;
+
   // Issue.  See you later in _enumerate_reply
   object_locator_t oloc(pool_id, ns);
-  pg_read(start.get_hash(), oloc, op,
-	  &on_ack->bl, 0, on_ack, &on_ack->epoch, &on_ack->budget);
+  // I hate having to do this. Try to find a cleaner way
+  // later.
+  pg_read(start.get_hash(), oloc, op, &on_ack->bl, 0,
+	  [c = std::move(on_ack)]
```

Comment:
seeing segfaults trying to access `&on_ack->bl` after moving out of `on_ack` (evaluation order of arguments isn't specified)

Reply:
cherry-picked